### PR TITLE
Replay pipeline using `cli exec` by downloading metadata

### LINKED ISSUE #1155: Allow to download pipline metadata via UI to replay with cli
I would be nice if I could download the metadata object via the UI as something like json to then use it to locally test with it (maybe adjusting some parts of it) via `cli exec`.

### PR DESCRIPTION
**Pull request by 6543**
closes  #1155

image
image

---
*Sponsored by Kithara Software GmbH*

### CODE REVIEW

File: woodpecker-go/woodpecker/pipeline.go [hunk 1]
```diff
@@ -11,3 +19,20 @@ func (c *client) PipelineQueue() ([]*Feed, error) {
 	err := c.get(uri, &out)
 	return out, err
 }
+
+// PipelineMetadata returns metadata for a pipeline, workflow name is optional.
+func (c *client) PipelineMetadata(repoID int64, pipelineNumber int, workflow ...string) ([]byte, error) {
```
Comment (6543) [inline woodpecker-go/woodpecker/pipeline.go:24]: my only two questions I'd like to have opinions from another persion:
- the golang lib: should we copy the metadata types into the client and deserialize or just handle it via []byte as it's done now ...
- should we use the _optional_ feature like done now or do some other users missinterprete this as list ...

**Comment by 6543** [comment]
image
image

more debug things could be added here later ... like to halt an pipeline and let exermine the curent state (e.g. dont cleanup the workspace) etc ...

**Comment by 6543** [comment]
second screenshot is with tab klicked .. it crrently only contains an dropdown and an button.

If you know a beter UI/UX please tell :) thats the best i did come up with

**Review by lafriks — CHANGES_REQUESTED** [review]
Secret values should not be returned, could be replaced with something like `<redacted>`

**Comment by 6543** [comment]
> Secret values should not be returned, could be replaced with something like `<redacted>`

there should not be any secrets in this metadata else we do have a problem

File: cmd/server/docs/docs.go [hunk 2]
```diff
@@ -5148,6 +5197,245 @@ const docTemplate = `{
...
+                "mask": {
+                    "type": "boolean"
+                },
+                "mount": {
+                    "type": "string"
+                },
+                "name": {
+                    "type": "string"
+                },
+                "value": {
```
Comment (lafriks) [inline cmd/server/docs/docs.go:5386]: What is this secret value used for?

Comment (6543) [reply cmd/server/docs/docs.go]: @lafriks good question!!!

that's legacy :dagger: 

8cf8aff3da24587d8cd18e51287c5927e174a1fc

**Review by lafriks — APPROVED** [review]

File: web/src/views/repo/pipeline/PipelineDebug.vue [hunk 3]
```diff
@@ -0,0 +1,76 @@
+<template>
+  <div v-if="repoPermissions && repoPermissions.push" class="p-4">
```
Comment (anbraten) [inline web/src/views/repo/pipeline/PipelineDebug.vue:2]: Could you  add a `Panel` here.

File: web/src/views/repo/pipeline/PipelineDebug.vue [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    <div class="flex items-center space-x-4">
+      <select
```
Comment (anbraten) [inline web/src/views/repo/pipeline/PipelineDebug.vue:4]: We have a `SelectField` for selects.

File: web/src/views/repo/pipeline/PipelineDebug.vue [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        <option v-for="workflow in workflows" :key="workflow" :value="workflow">{{ workflow }}</option>
+      </select>
+      <Button :is-loading="isLoading" :text="$t('repo.pipeline.debug.download_metadata')" @click="downloadMetadata" />
+    </div>
+  </div>
+  <div v-else class="flex items-center justify-center h-full">
+    <div class="text-center p-8 bg-wp-control-error-100 rounded-lg shadow-lg">
+      <p class="text-2xl font-bold text-white">{{ $t('repo.pipeline.debug.no_permission') }}</p>
+    </div>
+  </div>
```
Comment (anbraten) [inline web/src/views/repo/pipeline/PipelineDebug.vue:18]: We should hide the tab if missing permissions or just hide the element.

**Comment by anbraten** [comment]
What about running multiple workflows. From a UX perspective I would actually prefer some kind of `Debug locally` button to download all workflows as one file opening that file and running it with `cli exec`. #3967 should allow us to correctly run multiple workflows with `depends_on`.

File: web/src/views/repo/pipeline/PipelineDebug.vue (hunk 5 above)
Comment (6543) [reply web/src/views/repo/pipeline/PipelineDebug.vue:18]: the tab is hidden ... if you explizite navigate to the url you will se it

**Comment by 6543** [comment]
@anbraten added your pull as todo comment into the cli exec code ... 040b77348bfb561b04050def36fb5cc066a0f6a7

File: web/src/views/repo/pipeline/PipelineDebug.vue (hunk 3 above)
Comment (6543) [reply web/src/views/repo/pipeline/PipelineDebug.vue:2]: could you explain?

**Comment by 6543** [comment]
> What about running multiple workflows. From a UX perspective I would actually prefer some kind of `Debug locally` button to download all workflows as one file opening that file and running it with `cli exec`. #3967 should allow us to correctly run multiple workflows with `depends_on`.

you only need a specific workflow if you wanna run a workflow localy 1:1 else the generic e.g. pipeline metadata is good enouth

**Comment by anbraten** [comment]
> you only need a specific workflow if you wanna run a workflow localy 1:1 else the generic e.g. pipeline metadata is good enouth

As we always run pipelines on the server we should allow the user to be able to do so on the `cli` as well. Should be quite simple to create a single json like `{ 'workflow-one': <metadata>, 'workflow-two': <metadata> }` and load them in the cli accordingly to the workflows.

**Comment by 6543** [comment]
> As we always run pipelines on the server

well technically we just orchestrate them and no there are cases witch we filter out some.

also the orchestration logic is currently within the queue implementation and we can not easily port or reuse that on the cli.

this pull does not try to touch that part this is out of scope ...

... beside the rest we differ between server and cli is in the server.stepBuilder witch you are currently refactoring...

I get that it distracts you have to choose what to downolad, in this case i would propose to just drop it and dont export the workflow metadata as it only contains name and number anyway ...

**Comment by 6543** [comment]
@anbraten removed the workflow param and dropdown

File: web/src/views/repo/pipeline/PipelineDebug.vue (hunk 3 above)
Comment (6543) [reply web/src/views/repo/pipeline/PipelineDebug.vue]: done

**Comment by 6543** [comment]
image

File: web/src/assets/locales/en.json [hunk 6]
```diff
@@ -246,7 +246,16 @@
       "show_errors": "Show errors",
       "we_got_some_errors": "Oh no, we got some errors!",
       "duration": "Pipeline duration",
-      "created": "Created: {created}"
+      "created": "Created: {created}",
+      "debug": {
+        "title": "Debug",
+        "download_metadata": "Download Metadata File",
```
Comment (anbraten) [inline web/src/assets/locales/en.json:252]:
```suggestion
        "download_metadata": "Download metadata",
```

File: web/src/assets/locales/en.json [hunk 7]
```diff
@@ -246,7 +246,16 @@
...
+        "error_fetching": "Error fetching metadata",
+        "download_success": "Metadata downloaded successfully",
+        "no_permission": "You don't have permission to access debug information.",
+        "example_exec_title": "Rerun pipeline locally:",
```
Comment (anbraten) [inline web/src/assets/locales/en.json:256]:
```suggestion
        "metadata_exec_title": "Rerun pipeline locally",
```

File: web/src/assets/locales/en.json [hunk 8]
```diff
@@ -246,7 +246,16 @@
...
+        "example_exec_desc": "Download the pipeline metadata and run the pipeline locally again in the same context to fix any issues."
```
Comment (anbraten) [inline web/src/assets/locales/en.json:257]:
```suggestion
        "metadata_exec_desc": "Download this pipeline's metadata to run it pipeline locally again. This allows you to troubleshoot issues and test changes before committing them."
```

File: web/src/assets/locales/en.json [hunk 9]
```diff
@@ -246,7 +246,16 @@
       "show_errors": "Show errors",
       "we_got_some_errors": "Oh no, we got some errors!",
       "duration": "Pipeline duration",
-      "created": "Created: {created}"
+      "created": "Created: {created}",
+      "debug": {
+        "title": "Debug",
+        "download_metadata": "Download Metadata File",
+        "error_fetching": "Error fetching metadata",
```
Comment (anbraten) [inline web/src/assets/locales/en.json:253]: To have consistent keys:

```suggestion
        "metadata_download_error": "Error downloading metadata",
```

File: web/src/assets/locales/en.json [hunk 10]
```diff
@@ -246,7 +246,16 @@
...
+        "download_success": "Metadata downloaded successfully",
```
Comment (anbraten) [inline web/src/assets/locales/en.json:254]:
```suggestion
        "metadata_download_successful": "Metadata downloaded successfully",
```

File: web/src/views/repo/pipeline/PipelineDebug.vue [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+const notifications = useNotifications();
+
+const repo = inject<Ref<Repo>>('repo');
+const pipeline = inject<Ref<Pipeline>>('pipeline');
+const repoPermissions = inject<Ref<RepoPermissions>>('repo-permissions');
+
+const isLoading = ref(false);
+
+const downloadFileName = `${repo?.value.full_name.replaceAll('/', '-')}-pipeline-${pipeline?.value.number}-metadata.json`;
+const cliExecWithMetadata = `# woodpecker-cli exec --metadata-file ${downloadFileName}`;
```
Comment (anbraten) [inline web/src/views/repo/pipeline/PipelineDebug.vue:42]: This needs to be a `computed` as repo can change its value.

File: web/src/components/user/UserCLIAndAPITab.vue [hunk 12]
```diff
@@ -60,7 +60,7 @@ const usageWithCurl = computed(() => {
   return usage;
 });
 
-const usageWithCli = `# woodpecker setup --server ${address}`;
+const usageWithCli = `# woodpecker-cli setup --server ${address}`;
```
Comment (anbraten) [inline web/src/components/user/UserCLIAndAPITab.vue:63]: Isn't the cli name `woodpecker`?

```suggestion
const usageWithCli = `# woodpecker setup --server ${address}`;
```

File: web/src/views/repo/pipeline/PipelineDebug.vue (hunk 11 above)
Comment (anbraten) [inline web/src/views/repo/pipeline/PipelineDebug.vue:42]: This needs to be a `computed` as repo / pipeline can change their values:

```suggestion
const metadataFileName = computed(() => `${repo?.value.full_name.replaceAll('/', '-')}-pipeline-${pipeline?.value.number}-metadata.json`;
const cliExecWithMetadata = computed(() => `# woodpecker-cli exec --metadata-file ${metadataFileName.value}`);
```

File: web/src/views/repo/pipeline/PipelineDebug.vue [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    const metadata = await apiClient.getPipelineMetadata(repo.value.id, pipeline.value.number);
+
+    // Create a Blob with the JSON data
+    const blob = new Blob([JSON.stringify(metadata, null, 2)], { type: 'application/json' });
+
+    // Create a download link and trigger the download
+    const url = window.URL.createObjectURL(blob);
+    const link = document.createElement('a');
+    link.href = url;
+    link.download = downloadFileName;
```
Comment (6543) [inline web/src/views/repo/pipeline/PipelineDebug.vue:61]:
```suggestion
    link.download = metadataFileName;
```

File: web/src/views/repo/pipeline/PipelineDebug.vue [hunk 14]
```diff
@@ -0,0 +1,75 @@
+<template>
+  <template v-if="repoPermissions && repoPermissions.push">
+    <Panel>
+      <InputField :label="$t('repo.pipeline.debug.example_exec_title')">
+        <p class="text-sm text-wp-text-alt-100 mb-2">{{ $t('repo.pipeline.debug.example_exec_desc') }}</p>
```
Comment (6543) [inline web/src/views/repo/pipeline/PipelineDebug.vue:5]:
```suggestion
        <p class="text-sm text-wp-text-alt-100 mb-2">{{ $t('repo.pipeline.debug.metadata_exec_desc') }}</p>
```

File: web/src/views/repo/pipeline/PipelineDebug.vue [hunk 15]
```diff
@@ -0,0 +1,75 @@
+<template>
+  <template v-if="repoPermissions && repoPermissions.push">
+    <Panel>
+      <InputField :label="$t('repo.pipeline.debug.example_exec_title')">
```
Comment (6543) [inline web/src/views/repo/pipeline/PipelineDebug.vue:4]:
```suggestion
      <InputField :label="$t('repo.pipeline.debug.metadata_exec_title')">
```

File: web/src/views/repo/pipeline/PipelineDebug.vue [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    const url = window.URL.createObjectURL(blob);
+    const link = document.createElement('a');
+    link.href = url;
+    link.download = downloadFileName;
+    document.body.appendChild(link);
+    link.click();
+    document.body.removeChild(link);
+    window.URL.revokeObjectURL(url);
+
+    notifications.notify({ type: 'success', title: t('repo.pipeline.debug.download_success') });
```
Comment (6543) [inline web/src/views/repo/pipeline/PipelineDebug.vue:67]:
```suggestion
    notifications.notify({ type: 'success', title: t('repo.pipeline.debug.metadata_download_successful') });
```

File: web/src/views/repo/pipeline/PipelineDebug.vue (hunk 11 above)
Comment (6543) [reply web/src/views/repo/pipeline/PipelineDebug.vue]: applyed

**Comment by 6543** [comment]
@anbraten applyed

File: cli/exec/flags.go [hunk 17]
```diff
@@ -32,6 +32,11 @@ var flags = []cli.Flag{
 		Name:    "repo-path",
 		Usage:   "path to local repository",
 	},
+	&cli.StringFlag{
+		Sources: cli.EnvVars("WOODPECKER_METADATA_FILE"),
+		Name:    "metadata-file",
+		Usage:   "path to metadata file who has stored the pipeline environment to emulate (can be downloaded from an existing pipeline), infos can be overwritten.",
```
Comment (anbraten) [inline cli/exec/flags.go:38]:
```suggestion
		Usage:   "path to pipeline metadata file (normally downloaded from UI). Parameters can be adjusted by applying additional cli flags",
```

**Review by anbraten — APPROVED** [review]

**Pull request merged**